### DISCUSSION TITLE
Resource loading from classpath

### DISCUSSION THREAD

**Discussion by dayenda** [rfoltyns/log4j2-elasticsearch#65] [Q&A] [answered]
In the `ResourceUtil` class, classpath resource are loaded using `ClassLoader.getSystemClassLoader()`. This works fine when running from an IDE, but its giving me issues when running from a Spring Boot fat jar.

Is it possible to delegate resource loading to a child class loader (the spring boot class loader in my case), or could it make sense to use the current class loader instead of system class loader in `ResourceUtil`?

**Comment by rfoltyns** [comment, marked as answer]
What sort of issues are you having?
Where's the resource you're trying to load? Is it a resource of your Spring Boot app or is it a part of your app's dependencies?
Have you tried the solution used in [log4j2-elasticsearch-hc-springboot](https://github.com/rfoltyns/log4j2-elasticsearch-examples/blob/master/log4j2-elasticsearch-hc-springboot/src/main/resources/log4j2.xml)?
Could you post your configuration, please?

**Reply by dayenda** [reply]
I have have spring boot fat jar with my application classes and resource files like this:

app.jar
&nbsp;&nbsp;|- BOOT-INF
&nbsp;&nbsp;&nbsp;&nbsp;|- classes
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|- \<Resources and classes\>
&nbsp;&nbsp;&nbsp;&nbsp;|- lib
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|- dep.jar
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|- log4j2.yaml
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;|- log4j2-index-template.json

The log4j2.yaml file is referencing 'classpath:log4j2-index-template.json'. The log4j2.yaml configuration file is getting picked up, but i get an IllegalArgumentException in `ResourceUtil` because the referenced json file is not found using the system class loader.

When using a resource path like in the example you referenced (classpath:BOOT-INF/classes/log4j2-index-template.json) and placing the resource file in that folder it works. But since log4j2.yaml is getting picked up, I would expect to be able to keep my template files together with the configuration.

If I change `ResourceUtil.java` line 65 from `ClassLoader.getSystemClassLoader()`  to `LogAdapter.Log4jLog.class.getClassLoader()` it works as I would expect. I can see other places in the log4j code, that the classloader for the class `LogAdapter.Log4jLog` is used. 

I was wondering if it is the intention to use the system class loader in `ResourceUtil`?

Update:
The example with 'classpath:BOOT-INF/classes/log4j2-index-template.json' still works using `LogAdapter.Log4jLog.class.getClassLoader()`.

Thanks.

**Reply by rfoltyns** [reply]
In case of Spring Boot `ClassLoader.getSystemClassLoader()` forces the template to be packaged along with the application resources which sounds reasonable to me as serves several purposes:
* prevents dependencies from dictating how logging is configured
* ensures that resource can be loaded deterministically without jar loading order influencing the result in cases when more than one dependency provides the template

It can be relaxed - having a list of classloaders could be a potential improvement. But, `ClassLoader.getSystemClassLoader()` would still have priority, otherwise there would be no way to ensure 'one-source-of-truth' and override templates in dependencies. Backwards-compatibility and 'least surprise' are nice cherries on top here.

I can add it in 1.5.1. Would you be interested in creating a PR or would you rather leave it with me?

**Reply by dayenda** [reply]
I'll be happy to create a PR.

I understand your points, and it makes sense for me to use this differentiated resource loading with multiple class loaders. However, as an extension to log4j2, I would expect the resource loading to follow the same logic. A solution could be to simply delegate to `org.apache.logging.log4j.core.util.Loader::getResourceAsStream(String, ClassLoader)`. This would, however, mean that the system class loader would have the lowest priority. Alternatively, just delegate to the Loader class, after trying the system class loader. Which do you prefer?

**Reply by rfoltyns** [reply]
> as an extension to log4j2, I would expect the resource loading to follow the same logic

Even though it's an extension to Log4j2, it's only the `ElasticsearchAppender`, Log4j2 plugins and handful of other classes that should depend on it. Decoupling process is slowly happening since the last few months. As of 2.0, most of the code will be moved to [appenders](https://github.com/appenders) repos (including `ResourceUtil`). Appender, layout and plugins will be the only remaining classes.

> simply delegate to org.apache.logging.log4j.core.util.Loader::getResourceAsStream(String, ClassLoader)

Given the plans mentioned above, `ResourceUtil` cannot depend on any Log4j2 classes.

A small, private `ResourceUtil.getClassLoaders()` method with a list of system class loader and thread class loader should be enough here. `Find in classloaders -> break and load if found -> throw if not found`.

**Reply by dayenda** [reply]
I have created PR: https://github.com/rfoltyns/log4j2-elasticsearch/pull/67

**Reply by rfoltyns** [reply]
Thanks :+1:  Have a look at the changes I've added and let me know what you think.

**Reply by dayenda** [reply]
Thanks for your help 😊

**Reply by rfoltyns** [reply]
You're welcome! Happy to help! :slightly_smiling_face:
